parquet-variant/src/main/java/org/apache/parquet/variant/Variant.java

```diff
@@ -0,0 +1,288 @@
+  final ByteBuffer value;
+
+  /** The buffer that contains the Variant metadata. */
+  final ByteBuffer metadata;
+
+  /**
+   * The threshold to switch from linear search to binary search when looking up a field by key in
+   * an object. This is a performance optimization to avoid the overhead of binary search for a
+   * short list.
+   */
+  static final int BINARY_SEARCH_THRESHOLD = 32;
+
+  public Variant(byte[] value, byte[] metadata) {
+    this(value, 0, metadata, 0);
+  }
+
+  Variant(byte[] value, int valuePos, byte[] metadata, int metadataPos) {
+    this(
+        ByteBuffer.wrap(value, valuePos, value.length - valuePos),
+        ByteBuffer.wrap(metadata, metadataPos, metadata.length - metadataPos));
+  }
+
+  Variant(ByteBuffer value, ByteBuffer metadata) {
+    this.value = value.asReadOnlyBuffer();
+    this.value.mark();
+
+    this.metadata = metadata.asReadOnlyBuffer();
+    this.metadata.mark();
+
+    // There is currently only one allowed version.
+    if ((metadata.get(metadata.position()) & VariantUtil.VERSION_MASK) != VariantUtil.VERSION) {
+      throw new UnsupportedOperationException(String.format(
+          "Unsupported variant metadata version: %02X",
+          metadata.get(metadata.position()) & VariantUtil.VERSION_MASK));
+    }
+  }
+
+  /**
+   * @return the boolean value
+   */
+  public boolean getBoolean() {
+    return VariantUtil.getBoolean(value);
+  }
+
+  /**
+   * @return the byte value
+   */
+  public byte getByte() {
+    long longValue = VariantUtil.getLong(value);
+    if (longValue < Byte.MIN_VALUE || longValue > Byte.MAX_VALUE) {
+      throw new IllegalStateException("Value out of range for byte: " + longValue);
+    }
+    return (byte) longValue;
+  }
+
+  /**
+   * @return the short value
+   */
+  public short getShort() {
+    long longValue = VariantUtil.getLong(value);
+    if (longValue < Short.MIN_VALUE || longValue > Short.MAX_VALUE) {
+      throw new IllegalStateException("Value out of range for short: " + longValue);
+    }
+    return (short) longValue;
+  }
+
+  /**
+   * @return the int value
+   */
+  public int getInt() {
+    long longValue = VariantUtil.getLong(value);
+    if (longValue < Integer.MIN_VALUE || longValue > Integer.MAX_VALUE) {
+      throw new IllegalStateException("Value out of range for int: " + longValue);
+    }
+    return (int) longValue;
+  }
+
+  /**
+   * @return the long value
+   */
+  public long getLong() {
+    return VariantUtil.getLong(value);
+  }
+
+  /**
+   * @return the double value
+   */
+  public double getDouble() {
+    return VariantUtil.getDouble(value);
+  }
+
+  /**
+   * @return the decimal value
+   */
+  public BigDecimal getDecimal() {
+    return VariantUtil.getDecimal(value);
+  }
+
+  /**
+   * @return the float value
+   */
+  public float getFloat() {
+    return VariantUtil.getFloat(value);
+  }
+
+  /**
+   * @return the binary value
+   */
+  public byte[] getBinary() {
+    return VariantUtil.getBinary(value);
```

Comment:
Should this also return ByteBuffer? If not this requires a copy. If you used ByteBuffer then whether to copy is a decision made by the caller.